File: 02_ProgrammingFundamentals/_09_MapsLambdaAndStreamAPI/_02_Exercise/_09_ForceBook.java
package _02_ProgrammingFundamentals._09_MapsLambdaAndStreamAPI._02_Exercise;

import java.util.*;
import java.util.stream.Collectors;

public class _09_ForceBook {
    public static void main(String[] args) {
        Scanner scan = new Scanner(System.in);

        // Read input value
        String inputLine = scan.nextLine();

        // Add variables
        Map<String, String> users = new LinkedHashMap<>();
        Map<String, List<String>> forceAndUsers = new LinkedHashMap<>();

        // Create database for force users
        while (!inputLine.equals("Lumpawaroo")) {
            if (inputLine.contains("|")) {
                String[] line = inputLine.split(" \\| ");

                String side = line[0];
                String name = line[1];

                if (!users.containsKey(name) && !forceAndUsers.containsKey(side)) {
                    forceAndUsers.put(side, new ArrayList<>());
                    users.put(name, side);

                    forceAndUsers.get(side).add(name);

                } else if (!users.containsKey(name) && forceAndUsers.containsKey(side)) {
                    users.put(name, side);

                    forceAndUsers.get(side).add(name);

                } else if (users.containsKey(name)) {
                    inputLine = scan.nextLine();
                    continue;
                }

            } else if(inputLine.contains("->")) {
                String[] line = inputLine.split(" -> ");

                String name = line[0];
                String side = line[1];

                if (!users.containsKey(name) && !forceAndUsers.containsKey(side)) {
                    forceAndUsers.put(side, new ArrayList<>());
                    users.put(name, side);

                    forceAndUsers.get(side).add(name);

                    System.out.printf("%s joins the %s side!%n", name, side);

                } else if (!users.containsKey(name) && forceAndUsers.containsKey(side)) {
                    users.put(name, side);

                    forceAndUsers.get(side).add(name);

                    System.out.printf("%s joins the %s side!%n", name, side);

                } else if (users.containsKey(name) && forceAndUsers.containsKey(side)) {
                    forceAndUsers.get(users.get(name)).remove(name);
                    users.put(name, side);

                    forceAndUsers.get(side).add(name);

                    System.out.printf("%s joins the %s side!%n", name, side);

                } else if (users.containsKey(name) && !forceAndUsers.containsKey(side)) {
                    forceAndUsers.get(users.get(name)).remove(name);
                    users.put(name, side);

                    forceAndUsers.put(side, new ArrayList<>());
                    forceAndUsers.get(side).add(name);

                    System.out.printf("%s joins the %s side!%n", name, side);
                }
            }

            inputLine = scan.nextLine();
        }

        // Print result
        forceAndUsers.entrySet().stream().forEach(e -> {
            if(e.getValue().size() > 0){
                System.out.printf("Side: %s, Members: %d%n", e.getKey(), e.getValue().size());
                System.out.println(e.getValue().stream().map(element -> String.format("! %s", element)).collect(Collectors.joining("\n")));
            }
        });

        scan.close();
    }
}

File: 24_C#Fundamentals/_07_AssociativeArrays/_02_Exercise/_09_ForceBook/_09_ForceBook.cs
using System;
using System.Linq;
using System.Xml.Linq;

public class _09_ForceBook
{
    public static void Main(string[] args)
    {
        // Read input value
        string inputLine = Console.ReadLine();

        // Add variables
        Dictionary<string, string> users = new Dictionary<string, string>();
        Dictionary<string, List<string>> forceAndUsers = new Dictionary<string, List<string>>();

        // Create database for force users
        while (inputLine != "Lumpawaroo")
        {
            if (inputLine.Contains(" | "))
            {
                string[] line = inputLine.Split(" | ");

                string side = line[0];
                string name = line[1];

                if (!users.ContainsKey(name) && !forceAndUsers.ContainsKey(side))
                {
                    forceAndUsers[side] = new List<string>();
                    users[name] = side;

                    forceAndUsers[side].Add(name);

                }
                else if (!users.ContainsKey(name) && forceAndUsers.ContainsKey(side))
                {
                    users[name] = side;

                    forceAndUsers[side].Add(name);

                }
                else if (users.ContainsKey(name))
                {
                    inputLine = Console.ReadLine();
                    continue;
                }
            }
            else if (inputLine.Contains("->"))
            {
                string[] line = inputLine.Split(" -> ");

                string name = line[0];
                string side = line[1];

                if (!users.ContainsKey(name) && !forceAndUsers.ContainsKey(side))
                {
                    forceAndUsers[side] = new List<string>();
                    users[name] = side;

                    forceAndUsers[side].Add(name);

                    Console.WriteLine($"{name} joins the {side} side!");

                }
                else if (!users.ContainsKey(name) && forceAndUsers.ContainsKey(side))
                {
                    users[name] = side;

                    forceAndUsers[side].Add(name);

                    Console.WriteLine($"{name} joins the {side} side!");

                }
                else if (users.ContainsKey(name) && forceAndUsers.ContainsKey(side))
                {
                    forceAndUsers[users[name]].Remove(name);
                    users[name] = side;

                    forceAndUsers[side].Add(name);

                    Console.WriteLine($"{name} joins the {side} side!");

                }
                else if (users.ContainsKey(name) && !forceAndUsers.ContainsKey(side))
                {
                    forceAndUsers[users[name]].Remove(name);
                    users[name] = side;

                    forceAndUsers[side] = new List<string>();
                    forceAndUsers[side].Add(name);

                    Console.WriteLine($"{name} joins the {side} side!");
                }
            }

            inputLine = Console.ReadLine();
        }

        // Print result
        foreach (var entry in forceAndUsers)
        {
            if (entry.Value.Count > 0)
            {
                Console.WriteLine($"Side: {entry.Key}, Members: {entry.Value.Count}");
                Console.WriteLine(string.Join("\n", entry.Value.Select(element => $"! {element}")));
            }
        }
    }
}
